### PR TITLE
Update SEVIRI L2 BUFR reader to handle BUFR products from EUMETSAT Data Centre

### PR DESCRIPTION
**Pull request by ColinDuff**
This PR includes updates to the seviri l2 reader and yaml file so that MSG L2 BUFR products ordered from the UMARF cannow be read.

UMARF L2 products can contain a UMARF GS header or no header at all.

### CODE REVIEW

File: satpy/readers/seviri_l2_bufr.py [hunk 1]
```diff
@@ -19,13 +19,14 @@
 
 
 import logging
-from datetime import timedelta
+from datetime import timedelta, datetime
 import numpy as np
 import xarray as xr
 import dask.array as da
-
+from fnmatch import fnmatch
```
Comment (sjoro) [inline satpy/readers/seviri_l2_bufr.py]: module not needed, see comment on line 61

File: satpy/readers/seviri_l2_bufr.py [hunk 2]
```diff
@@ -19,13 +19,14 @@
...
 from satpy.readers.seviri_base import mpef_product_header
 from satpy.readers.eum_base import recarray2dict
+import os
```
Comment (sjoro) [inline satpy/readers/seviri_l2_bufr.py]: module not needed, see comment on line 61

File: satpy/readers/seviri_l2_bufr.py [hunk 3]
```diff
@@ -85,6 +100,22 @@ def _read_mpef_header(self):
         hdr = np.fromfile(self.filename, mpef_product_header, 1)
         return recarray2dict(hdr)
 
+    def get_attribute(self, key):
+        ''' Get BUFR attributes '''
+        fh = open(self.filename, "rb")
+        while 1:
```
[the remark this answers is not in the document]
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: I originally did it this way but it somehow causes a problem when i reopen the file to get the data . almost as if the next time i open the file it is starting from the next message. On my list of things to investigate. Maybe the close isnt working properly.

[the remark this answers is not in the document]
Comment (sjoro) [reply satpy/readers/seviri_l2_bufr.py]: maybe you can try `fh.seek(0)` before closing?

[the remark this answers is not in the document]
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: will try that , cheers

File: satpy/readers/seviri_l2_bufr.py [hunk 4]
```diff
@@ -54,8 +56,21 @@ def __init__(self, filename, filename_info, filetype_info, **kwargs):
                                                       filename_info,
                                                       filetype_info)
 
-        self.filename = filename
-        self.mpef_header = self._read_mpef_header()
+        # Offline filenames are in the following naming convention
+        # GIIBUFRProduct_20190515151500Z_00_OMPEFS04_MET11_FES_E0000
+        if (fnmatch(os.path.basename(filename), '*BUFRProd*MPEFS*')):
```
[the remark this answers is not in the document]
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: done

File: satpy/readers/seviri_l2_bufr.py [hunk 5]
```diff
@@ -38,8 +39,9 @@
 
 logger = logging.getLogger('BufrProductClasses')
 
+umarf_dict = {55: {'ssp': 'E0415', 'name': '08'}, 56:  {'ssp': 'E0000', 'name': '09'},
```
[the remark this answers is not in the document]
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: done

File: satpy/readers/seviri_l2_bufr.py [hunk 6]
```diff
@@ -54,8 +56,21 @@ def __init__(self, filename, filename_info, filetype_info, **kwargs):
...
 
-        self.filename = filename
-        self.mpef_header = self._read_mpef_header()
+        # Offline filenames are in the following naming convention
+        # GIIBUFRProduct_20190515151500Z_00_OMPEFS04_MET11_FES_E0000
+        if (fnmatch(os.path.basename(filename), '*BUFRProd*MPEFS*')):
+            # EUMETSAT Offline Bufr product
+            self.mpef_header = self._read_mpef_header()
+        else:
+            # Product was retrieved from the UMARF
```
[the remark this answers is not in the document]
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: done

File: satpy/readers/seviri_l2_bufr.py (hunk 1 above)
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: removed

File: satpy/readers/seviri_l2_bufr.py (hunk 2 above)
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: removed

File: satpy/readers/seviri_l2_bufr.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: cant seem to resolve this issue - have raised an issue on the python-eccodes github page. can we leave it as is for now.

[the remark this answers is not in the document]
Comment (sjoro) [reply satpy/readers/seviri_l2_bufr.py]: i'm ok with this as this is not a bug. should be revisited if/when more info is received from ecCodes.

[the remark this answers is not in the document]
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: as discussed an Issue can be raised for this, comments also in the code explaining the problem. Conversation can now be resolved.

**Comment by ColinDuff** [comment]
Please note UMARF is the EUMETSAT data centre. The BUFR products from there come with or without Headers - neither come with the Mpef Product Header.

File: satpy/readers/seviri_l2_bufr.py [hunk 7]
```diff
@@ -54,8 +54,19 @@ def __init__(self, filename, filename_info, filetype_info, **kwargs):
...
+            self.mpef_header = self._read_mpef_header()
+        else:
+            # Product was retrieved from the EUMETSAT Data Centre
+            timeStr = self.get_attribute('typicalDate')+self.get_attribute('typicalTime')
+            buf_start_time = datetime.strptime(timeStr, "%Y%m%d%H%M%S")
+            sc_id = self.get_attribute('satelliteIdentifier')
+            self.mpef_header = {}
+            self.mpef_header['NominalTime'] = buf_start_time
+            self.mpef_header['SpacecraftName'] = data_centre_dict[sc_id]['name']
+            self.mpef_header['RectificationLongitude'] = data_centre_dict[sc_id]['ssp']
```
Comment (mraspaud) [inline satpy/readers/seviri_l2_bufr.py]: Should this be passed to the dataset array as metadata ?

File: satpy/readers/seviri_l2_bufr.py [hunk 8]
```diff
@@ -38,8 +37,9 @@
 
 logger = logging.getLogger('BufrProductClasses')
 
+data_centre_dict = {55: {'ssp': 'E0415', 'name': '08'}, 56:  {'ssp': 'E0000', 'name': '09'},
```
Comment (mraspaud) [inline satpy/readers/seviri_l2_bufr.py]: The official language of satpy is American English, so it should be 'center' (sorry...)

File: satpy/readers/seviri_l2_bufr.py [hunk 9]
```diff
@@ -85,6 +96,24 @@ def _read_mpef_header(self):
         hdr = np.fromfile(self.filename, mpef_product_header, 1)
         return recarray2dict(hdr)
 
+    def get_attribute(self, key):
+        ''' Get BUFR attributes '''
+        # This function is inefficient as it is looping through the entire
+        # file to get 1 attribute. It causes a problem though if you break
+        # from the file early - dont knowgit  why but investigating - fix later
+        fh = open(self.filename, "rb")
+        while 1:
```
Comment (mraspaud) [inline satpy/readers/seviri_l2_bufr.py]:
```suggestion
        while True:
```
More pythonic...

File: satpy/readers/seviri_l2_bufr.py (hunk 7 above)
Comment (mraspaud) [reply satpy/readers/seviri_l2_bufr.py]: Or is it already ?

Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: spacecraftname and Rectification and set as metadata already 

xarr.attrs['spacecraft_name'] = self.spacecraft_name
xarr.attrs['ssp_lon'] = self.ssp_lon

start time is a scene attribute so i guess that is also ok - Sauli?

Comment (sjoro) [reply satpy/readers/seviri_l2_bufr.py]: looks good to me.

File: satpy/readers/seviri_l2_bufr.py [hunk 10]
```diff
@@ -98,6 +98,9 @@ def _read_mpef_header(self):
 
     def get_attribute(self, key):
         ''' Get BUFR attributes '''
+        # This function is inefficient as it is looping through the entire
+        # file to get 1 attribute. It causes a problem though if you break
+        # from the file early - dont knowgit  why but investigating - fix later
```
Comment (sjoro) [inline satpy/readers/seviri_l2_bufr.py]: typo "knowgit"?!?, please fix to clarify.

File: satpy/readers/seviri_l2_bufr.py (hunk 7 above)
Comment (mraspaud) [reply satpy/readers/seviri_l2_bufr.py]: Perfect then

File: satpy/readers/seviri_l2_bufr.py (hunk 8 above)
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: fixed

File: satpy/readers/seviri_l2_bufr.py [hunk 11]
```diff
@@ -38,8 +37,9 @@
 
 logger = logging.getLogger('BufrProductClasses')
```
[the remark this answers is not in the document]
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: done

File: satpy/readers/seviri_l2_bufr.py (hunk 10 above)
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: fixed

File: satpy/readers/seviri_l2_bufr.py (hunk 9 above)
Comment (ColinDuff) [reply satpy/readers/seviri_l2_bufr.py]: fixed

**Comment by mraspaud** [comment]
@ColinDuff @sjoro there seems be a conflict with master now, could you have a look at it ? After that I think it's ready to merge.

**Comment by sjoro** [comment]
this might be related to my PR [#1041](https://github.com/pytroll/satpy/pull/1041). @ColinDuff, can you please have a look so we get this PR merged.

**Comment by ColinDuff** [comment]
conflict resolved, should be ok now

**Comment by mraspaud** [comment]
Thanks, merging now.

**Pull request merged**